gym/wrappers/time_limit.py

```diff
@@ -1,44 +1,24 @@
-import time
-from gym import Wrapper, logger
+import gym
 
-class TimeLimit(Wrapper):
-    def __init__(self, env, max_episode_seconds=None, max_episode_steps=None):
+
+class TimeLimit(gym.Wrapper):
+    def __init__(self, env, max_episode_steps=None):
         super(TimeLimit, self).__init__(env)
-        self._max_episode_seconds = max_episode_seconds
+        if max_episode_steps is None:
+            max_episode_steps = env.spec.max_episode_steps
+        self.env.spec.max_episode_steps = max_episode_steps
         self._max_episode_steps = max_episode_steps
-
-        self._elapsed_steps = 0
-        self._episode_started_at = None
-
-    @property
-    def _elapsed_seconds(self):
-        return time.time() - self._episode_started_at
-
-    def _past_limit(self):
-        """Return true if we are past our limit"""
-        if self._max_episode_steps is not None and self._max_episode_steps <= self._elapsed_steps:
-            logger.debug("Env has passed the step limit defined by TimeLimit.")
-            return True
-
-        if self._max_episode_seconds is not None and self._max_episode_seconds <= self._elapsed_seconds:
-            logger.debug("Env has passed the seconds limit defined by TimeLimit.")
-            return True
-
-        return False
+        self._elapsed_steps = None
 
     def step(self, action):
-        assert self._episode_started_at is not None, "Cannot call env.step() before calling reset()"
+        assert self._elapsed_steps is not None, "Cannot call env.step() before calling reset()"
         observation, reward, done, info = self.env.step(action)
         self._elapsed_steps += 1
-
-        if self._past_limit():
-            if self.metadata.get('semantics.autoreset'):
-                _ = self.reset() # automatically reset the env
-            done = True 
-
+        if self._elapsed_steps >= self._max_episode_steps:
+            info['TimeLimit.truncated'] = not done
+            done = True
```

Comment:
I see, but I guess this gonna break a lot of existing code in RL community, particularly collecting dat is heavily reply on done to break the for loop for instance.

Reply:
This will indeed break code which previously relied on the `done` flag to decide if a time-out occurred. Although possibly disruptive to some users, I would argue that that's the direction we should take here, since any code that does that is already dealing with time-outs in the wrong way (conflating time-outs with reaching terminal states), and should likely be changed.